### PR TITLE
Ia 70/create celery task for collect statistic

### PR DESCRIPTION
**Pull request by flying-pi**

### CODE REVIEW

File: rg_instructor_analytics/models.py [hunk 1]
```diff
@@ -42,3 +43,30 @@ class Meta:
...
         unique_together = ('course_id', 'student',)
+
+
+class GradeStatistic(Model):
+    """
+    Model for store grades of the student for each course.
+    """
+
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    student = ForeignKey(User, db_index=True)
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: the default value is not required to the filed definition

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: rg_instructor_analytics/models.py [hunk 2]
```diff
@@ -42,3 +43,30 @@ class Meta:
...
+
+    class Meta:
+        """
+        Meta class.
+        """
+
+        unique_together = ('course_id', 'student',)
+
+
+# TODO replace this table with redis.
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: typo: `the Redis`

File: rg_instructor_analytics/models.py [hunk 3]
```diff
@@ -42,3 +43,30 @@ class Meta:
...
+class LastGradeStatUpdate(Model):
+    """
+    Used for calculate diffs for update.
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: typo: s/Used for calculate diffs for update./For the calculation diffs for the update./

File: rg_instructor_analytics/tasks.py [hunk 4]
```diff
@@ -1,28 +1,44 @@
...
 from lms import CELERY_APP
 from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
-from rg_instructor_analytics.models import EnrollmentByStudent, EnrollmentTabCache
+from rg_instructor_analytics.models import EnrollmentByStudent, EnrollmentTabCache, GradeStatistic, LastGradeStatUpdate
+from student.models import CourseEnrollment
+from xmodule.modulestore.django import modulestore
+
+try:
+    from lms.djangoapps.grades.new.course_grade_factory import CourseGradeFactory
+except Exception:
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: Please use an exact exception.

File: rg_instructor_analytics/tasks.py [hunk 5]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
             )
 
     cache.set(ENROLLMENT_STAT_CACHE_BY_COURSE_KEY, total_stat)
+
+
+def get_items_for_grade_update():
+    """
+    Return aggregate list of the users by course, those grades need be recalculated.
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: typo: s/Return aggregate list of the users by course, those grades need be recalculated./Returns an aggregate list of the users by the course, those grades need to be recalculated./

File: rg_instructor_analytics/tasks.py [hunk 6]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
...
+    """
+    last_update_info = LastGradeStatUpdate.objects.all()
+    if last_update_info.exists():
+        item_for_update = (
+            StudentModule.objects
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: Please add comments why requests are done on different models.

File: rg_instructor_analytics/tasks.py [hunk 7]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
...
     cache.set(ENROLLMENT_STAT_CACHE_BY_COURSE_KEY, total_stat)
+
+
+def get_items_for_grade_update():
+    """
+    Return aggregate list of the users by course, those grades need be recalculated.
+    """
+    last_update_info = LastGradeStatUpdate.objects.all()
+    if last_update_info.exists():
+        item_for_update = (
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: Please use correct name to the variable if an object is the plural use plural name.

File: rg_instructor_analytics/tasks.py [hunk 8]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
...
+    for item in item_for_update:
+        if item['course_id'] not in users_by_course:
+            users_by_course[item['course_id']] = []
+        users_by_course[item['course_id']].append(item['student__id'])
+    return users_by_course
+
+
+def get_grade_summary(user_id, course):
+    """
+    Return grade of the given user for the given course.
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: typo: s/Return grade of the given user for the given course./Returns the grade for the given student in the addressed course.//

File: rg_instructor_analytics/tasks.py [hunk 9]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
...
+                exam_info = OrderedDict()
+                for grade in grades['section_breakdown']:
+                    exam_info[grade['label']] = int(grade['percent'] * 100.0)
+                total = int(grades['percent'] * 100.0)
+                exam_info['total'] = total
+
+                collected_stat.append(
+                    (
+                        {'course_id': course_key, 'student_id': user},
+                        {'exam_info': json.dumps(exam_info), 'total': int(grades['percent'] * 100.0)}
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: please try to avoid code duplication

File: rg_instructor_analytics/tasks.py [hunk 10]
```diff
@@ -118,3 +134,90 @@ def enrollment_collector_date():
...
+                    )
+                )
+
+    with transaction.atomic():
+        for key_values, additiona_info in collected_stat:
+            key_values['defaults'] = additiona_info
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: what does `additiona` mean?

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: rg_instructor_analytics/models.py (hunk 2 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done

File: rg_instructor_analytics/models.py (hunk 3 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done

File: rg_instructor_analytics/tasks.py (hunk 4 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 5 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 6 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 7 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 8 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 9 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: Thank you, fixed.

File: rg_instructor_analytics/tasks.py (hunk 10 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: fixed

**Review by idegtiarov — APPROVED** [review]

**Pull request merged**